### PR TITLE
IGAPP-619: Sanitize remote content on web

### PR DESCRIPTION
**Pull request by klinzo**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

### CODE REVIEW

File: web/src/components/RemoteContent.tsx [hunk 1]
```diff
@@ -59,68 +60,46 @@ const SandBox = styled.div<{ centered: boolean }>`
...
       }
     })
-  }
+  }, [html, handleClick, sandBoxRef])
 
-  componentDidMount(): void {
-    this.hijackATags()
+  const dangerouslySetInnerHTML = {
+    __html: Dompurify.sanitize(html)
   }
```
Comment (klinzo) [inline web/src/components/RemoteContent.tsx:100]: The important change in this PR :)

**Comment by klinzo** [comment]
Another try at sanitizing our html. Not sure how we should proceed with native and the react-native-webview.

File: web/src/components/RemoteContent.tsx [hunk 2]
```diff
@@ -59,68 +60,46 @@ const SandBox = styled.div<{ centered: boolean }>`
...
 const HIJACK = new RegExp(buildConfig().internalLinksHijackPattern)
 
-class RemoteContent extends React.Component<PropsType> {
-  static defaultProps = {
-    centered: false
-  }
-
-  sandBoxRef: RefObject<HTMLDivElement>
+const RemoteContent = ({ html, onInternalLinkClick, centered = false }: PropsType): ReactElement => {
+  const sandBoxRef: RefObject<HTMLDivElement> = React.createRef<HTMLDivElement>()
```
Comment (Taggotty) [inline web/src/components/RemoteContent.tsx:71]:
```suggestion
  const sandBoxRef = React.createRef<HTMLDivElement>()
```

File: web/src/components/RemoteContent.tsx [hunk 3]
```diff
@@ -59,68 +60,46 @@ const SandBox = styled.div<{ centered: boolean }>`
...
-
-    if (target instanceof HTMLAnchorElement) {
-      const href = target.href
-      this.props.onInternalLinkClick(decodeURIComponent(new URL(decodeURIComponent(href)).pathname))
-    }
-  }
+  const handleClick = useCallback(
+    (event: MouseEvent): void => {
+      event.preventDefault ? event.preventDefault() : (event.returnValue = false)
+      const target: EventTarget | null = event.currentTarget
```
Comment (Taggotty) [inline web/src/components/RemoteContent.tsx:76]:
```suggestion
      const target = event.currentTarget
```

File: web/src/components/RemoteContent.tsx [hunk 4]
```diff
@@ -59,68 +60,46 @@ const SandBox = styled.div<{ centered: boolean }>`
...
+  )
 
-  hijackATags(): void {
-    if (!this.sandBoxRef.current) {
+  useEffect(() => {
+    if (!sandBoxRef.current) {
       return
     }
-    const collection: HTMLCollectionOf<HTMLAnchorElement> = this.sandBoxRef.current.getElementsByTagName('a')
+    const collection: HTMLCollectionOf<HTMLAnchorElement> = sandBoxRef.current.getElementsByTagName('a')
```
Comment (Taggotty) [inline web/src/components/RemoteContent.tsx:90]:
```suggestion
    const collection = sandBoxRef.current.getElementsByTagName('a')
```

File: web/src/components/RemoteContent.tsx (hunk 1 above)
Comment (Taggotty) [reply web/src/components/RemoteContent.tsx:100]: So I guess this means we discarded the idea to sanitize the content in the api client? Or do I mix this up with the news?

**Review by Taggotty — APPROVED** [review]

File: api-client/src/endpoints/__tests__/createDisclaimerEndpoint.spec.ts [hunk 5]
```diff
@@ -9,7 +10,7 @@ describe('disclaimer', () => {
     title: 'Feedback, Kontakt und mögliches Engagement',
     type: 'disclaimer',
     modified_gmt: '2017-06-12 12:27:57',
-    content: '<a href="javascript:IWantToBeRemoved();">Ich bleib aber da.</a>',
+    content: '<div>Some disclaimer test content :)</div>',
```
Comment (maxammann) [inline api-client/src/endpoints/__tests__/createDisclaimerEndpoint.spec.ts:13]: I think removing the javascript makes the test invalid. I think this test is now only testing that the content is not changing.

File: api-client/src/endpoints/__tests__/createEventsEndpoint.spec.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ describe('events', () => {
     path: '/augsburg/de/events/asylpolitischer_fruehschoppen',
     title: 'Asylpolitischer Frühschoppen',
     excerpt: 'Asylpolitischer Fr&uuml;hschoppen',
-    content: '<a href="javascript:IWantToBeRemoved();">Ich bleib aber da.</a>',
+    content: '<div>Some event test content :)</div>',
```
Comment (maxammann) [inline api-client/src/endpoints/__tests__/createEventsEndpoint.spec.ts:28]: Same here, I think this test is not correct anymore.

File: web/src/components/__tests__/RemoteContent.spec.tsx [hunk 7]
```diff
@@ -50,4 +50,22 @@ describe('RemoteContent', () => {
 
     expect(onInternalLinkClick).toHaveBeenCalledTimes(0)
   })
+
+  it('should sanitize the html', () => {
```
Comment (maxammann) [inline web/src/components/__tests__/RemoteContent.spec.tsx:54]: Did you test whether this test really fails if you remove the dom purify sanitization?

**Review by maxammann — APPROVED** [review]
If I understood this PR correctly then we moved sanitization from the api-client to the rendering. This is good! Was the previous PR rejected?

What's missing is the same sanitization for the react native webview. But I think we have a task for that.

File: web/src/components/RemoteContent.tsx (hunk 1 above)
Comment (klinzo) [reply web/src/components/RemoteContent.tsx:100]: Yes, the problem was that neither dompurify nor isomorphic-dompurify work on native (isomorphic-dompurify only works in browser and node). Also imo it is better to sanitize where it is potentially dangerous, i.e. when setting it as inner html. That way, we make sure everything is really sanitized and safe.

**Comment by klinzo** [comment]
> If I understood this PR correctly then we moved sanitization from the api-client to the rendering. This is good! Was the previous PR rejected?
> 
> What's missing is the same sanitization for the react native webview. But I think we have a task for that.

@maxammann no, it was merged but as neither dompurify nor isomorphic-dompurify did work with react-native I had to revert it: #345 

I'll create an issue for that :)

File: web/src/components/__tests__/RemoteContent.spec.tsx (hunk 7 above)
Comment (klinzo) [reply web/src/components/__tests__/RemoteContent.spec.tsx:54]: Adjusted it.

**Review by Taggotty — APPROVED** [review]

**Pull request merged**